kratos/utilities/auxiliar_model_part_utilities.h

```diff
@@ -747,6 +850,110 @@ class KRATOS_API(KRATOS_CORE) AuxiliarModelPartUtilities
+    void AuxiliaryAddEntitiesWithNodes(
+        TEntitiesContainer& rEntitiesContainer,
+        const std::vector<IndexType>& rEntitiesIds,
+        IndexType ThisIndex = 0
+        )
+    {
+        KRATOS_TRY
+        
+        // Obtain from the root model part the corresponding list of nodes
+        const auto it_ent_end = rEntitiesContainer.end();
+        std::unordered_set<IndexType> set_of_node_ids;
+        for(IndexType i=0; i<rEntitiesIds.size(); ++i) {
+          auto it_ent = rEntitiesContainer.find(rEntitiesIds[i]);
+          if(it_ent!=it_ent_end) {
+            const auto& r_geom = it_ent->GetGeometry();
+            for (IndexType j = 0; j < r_geom.size(); ++j) {
+              set_of_node_ids.insert(r_geom[j].Id());
+            }
+          } else {
+            KRATOS_ERROR << "The entity with Id " << rEntitiesIds[i] << " does not exist in the root model part";
+          }
+        }
+
+        // Adding nodes
+        std::vector<IndexType> list_of_nodes;
+        list_of_nodes.insert(list_of_nodes.end(), set_of_node_ids.begin(), set_of_node_ids.end());
+        mrModelPart.AddNodes(list_of_nodes);
+
+        // Add to all of the leaves
+        ModelPart* p_current_part = &mrModelPart;
+        while(p_current_part->IsSubModelPart()) {
+          p_current_part->AddNodes(list_of_nodes);
+          p_current_part = &(p_current_part->GetParentModelPart());
+        }
+
+        KRATOS_CATCH("")
+    }
+
+    /**
+     * @brief Inserts a list of pointers to elements and the belonging nodes
+     * @param rEntitiesContainer The entities to be added
+     * @param ItElementsBegin The begin iterator
+     * @param ItElementsEnd The end iterator
+     * @param ThisIndex The mesh index
+     * @tparam TEntitiesContainer The class of entities considered
+     * @tparam TIteratorType The class of iterator considered
+     */
+    template<class TEntitiesContainer, class TIteratorType>
+    void AuxiliaryAddEntitiesWithNodes(
+        TEntitiesContainer& rEntitiesContainer,
+        TEntitiesContainer& rAux,
+        std::vector<IndexType>& rListOfNodes,
+        TIteratorType ItEntitiesBegin,
+        TIteratorType ItEntitiesEnd,
+        IndexType ThisIndex = 0
+        )
+    {
+        KRATOS_TRY
+        
+        TEntitiesContainer aux_root;
+        std::unordered_set<IndexType> set_of_nodes;
+        
+        const auto it_ent_end = rEntitiesContainer.end();
+        for(TIteratorType it_ent = ItEntitiesBegin; it_ent!=ItEntitiesEnd; ++it_ent) {
+            auto it_ent_found = rEntitiesContainer.find(it_ent->Id());
+            if(it_ent_found == it_ent_end) { // Entity does not exist in the top model part
+                aux_root.push_back( *(it_ent.base()) );
+                rAux.push_back( *(it_ent.base()) );
+                const auto& r_geom = it_ent->GetGeometry();
+                for (IndexType i = 0; i < r_geom.size(); ++i) {
+                    set_of_nodes.insert(r_geom[i].Id());
```

Comment:
I would put this repeated code in a separate private method.

Reply:
But this code is on a templated function, I don't know what is required to refactor or do you mean other pieces of code?